File: clients/android/app/src/main/java/asia/tnestai/vpn/UsageLedger.kt
package asia.tnestai.vpn

import android.content.Context
import java.time.Instant
import java.time.ZoneId

class UsageLedger(context: Context) {
    data class Totals(
        val sessionDownload: Long,
        val sessionUpload: Long,
        val todayDownload: Long,
        val todayUpload: Long,
        val monthDownload: Long,
        val monthUpload: Long
    )

    private val prefs = context.getSharedPreferences("usage", Context.MODE_PRIVATE)
    private val zone = ZoneId.of("Asia/Shanghai")
    private var lastRx = -1L
    private var lastTx = -1L
    private var sessionRx = 0L
    private var sessionTx = 0L
    private var day = prefs.getString("day", "") ?: ""
    private var month = prefs.getString("month", "") ?: ""
    private var dayRx = prefs.getLong("day_rx", 0)
    private var dayTx = prefs.getLong("day_tx", 0)
    private var monthRx = prefs.getLong("month_rx", 0)
    private var monthTx = prefs.getLong("month_tx", 0)
    private var lastPersist = 0L

    fun update(rx: Long, tx: Long, nowMillis: Long = System.currentTimeMillis()): Totals {
        val local = Instant.ofEpochMilli(nowMillis).atZone(zone)
        val nextDay = local.toLocalDate().toString()
        val nextMonth = "${local.year}-${local.monthValue.toString().padStart(2, '0')}"
        if (day != nextDay) {
            day = nextDay
            dayRx = 0
            dayTx = 0
        }
        if (month != nextMonth) {
            month = nextMonth
            monthRx = 0
            monthTx = 0
        }
        if (lastRx >= 0) {
            val deltaRx = if (rx >= lastRx) rx - lastRx else rx
            val deltaTx = if (tx >= lastTx) tx - lastTx else tx
            sessionRx += deltaRx
            sessionTx += deltaTx
            dayRx += deltaRx
            dayTx += deltaTx
            monthRx += deltaRx
            monthTx += deltaTx
        }
        lastRx = rx
        lastTx = tx
        if (nowMillis - lastPersist >= 30_000) {
            prefs.edit()
                .putString("day", day).putString("month", month)
                .putLong("day_rx", dayRx).putLong("day_tx", dayTx)
                .putLong("month_rx", monthRx).putLong("month_tx", monthTx)
                .apply()
            lastPersist = nowMillis
        }
        return Totals(sessionRx, sessionTx, dayRx, dayTx, monthRx, monthTx)
    }
}

File: clients/windows/src/TNestVPN.Core/UsageLedger.cs
using System.Text.Json;

namespace TNestVPN.Core;

public sealed record UsageTotals(
    long SessionDownload, long SessionUpload,
    long TodayDownload, long TodayUpload,
    long MonthDownload, long MonthUpload);

public sealed class UsageLedger
{
    private sealed class State
    {
        public string Day { get; set; } = "";
        public string Month { get; set; } = "";
        public long DayRx { get; set; }
        public long DayTx { get; set; }
        public long MonthRx { get; set; }
        public long MonthTx { get; set; }
    }

    private readonly string path = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "TNest VPN", "usage.json");
    private State state;
    private long lastRx = -1, lastTx = -1, sessionRx, sessionTx;
    private DateTimeOffset lastPersist = DateTimeOffset.MinValue;

    public UsageLedger()
    {
        try { state = JsonSerializer.Deserialize<State>(File.ReadAllText(path)) ?? new(); }
        catch { state = new(); }
    }

    public UsageTotals Update(long rx, long tx, DateTimeOffset now)
    {
        var shanghai = TimeZoneInfo.ConvertTime(now,
            TimeZoneInfo.FindSystemTimeZoneById("China Standard Time"));
        var day = shanghai.ToString("yyyy-MM-dd");
        var month = shanghai.ToString("yyyy-MM");
        if (state.Day != day) {
            state.Day = day;
            state.DayRx = state.DayTx = 0;
        }
        if (state.Month != month) {
            state.Month = month;
            state.MonthRx = state.MonthTx = 0;
        }
        if (lastRx >= 0) {
            var deltaRx = rx >= lastRx ? rx - lastRx : rx;
            var deltaTx = tx >= lastTx ? tx - lastTx : tx;
            sessionRx += deltaRx;
            sessionTx += deltaTx;
            state.DayRx += deltaRx;
            state.DayTx += deltaTx;
            state.MonthRx += deltaRx;
            state.MonthTx += deltaTx;
        }
        lastRx = rx;
        lastTx = tx;
        if (now - lastPersist >= TimeSpan.FromSeconds(30)) {
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);
            File.WriteAllText(path, JsonSerializer.Serialize(state));
            lastPersist = now;
        }
        return new(sessionRx, sessionTx, state.DayRx, state.DayTx,
            state.MonthRx, state.MonthTx);
    }
}
